src/thunderbots/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,272 @@
+        std::this_thread::yield();
+
+        // Take ownership of the `in_destructor` flag so we can use it for the conditional
+        // check
+        in_destructor_mutex.lock();
+    }
+}
+
+void PassGenerator::optimizePasses()
+{
+    // The objective function we minimize in gradient descent to improve each pass
+    // that we're optimizing
+    const auto objective_function =
+        [this](std::array<double, NUM_PARAMS_TO_OPTIMIZE> pass_array) {
+            Pass pass = convertArrayToPass(pass_array);
+            return ratePass(pass);
+        };
+
+    // Run gradient descent to optimize the passes to for the requested number
+    // of iterations
+    // NOTE: Parallelizing this `for` loop would probably be safe and potentially more
+    //       performant
+    for (Pass& pass : passes_to_optimize)
+    {
+        auto pass_array =
+            optimizer.maximize(objective_function, convertPassToArray(pass),
+                               number_of_gradient_descent_steps_per_iter.value());
+        pass = convertArrayToPass(pass_array);
+    }
+}
+
+void PassGenerator::pruneAndReplacePasses()
+{
+    // Sort the passes by decreasing quality
+    std::sort(passes_to_optimize.begin(), passes_to_optimize.end(),
+              [this](Pass p1, Pass p2) { return comparePassQuality(p1, p2); });
+
+    // Merge Passes That Are Similar
+    // We start by assuming that the most similar passes will be right beside each other,
+    // then iterate over the entire list, building a new list as we go by only adding
+    // elements when they are dissimilar enough from the last element we added
+    passes_to_optimize = std::accumulate(
+        passes_to_optimize.begin(), passes_to_optimize.end(), std::vector<Pass>(),
+        [this](std::vector<Pass>& passes, Pass curr_pass) {
+            // Check if we have no passes, or if this pass is too similar to the
+            // last pass we added to the list
+            if (passes.empty() || !passesEqual(curr_pass, passes.back()))
+            {
+                passes.emplace_back(curr_pass);
+            }
+            return passes;
+        });
+
+    // Replace the least promising passes with newly generated passes
+    if (num_passes_to_keep_after_pruning.value() < num_passes_to_optimize.value())
+    {
+        std::vector<Pass> new_passes = generatePasses(
+            num_passes_to_optimize.value() - num_passes_to_keep_after_pruning.value());
+
+        // Remove the worst paths
+        if (num_passes_to_keep_after_pruning.value() < passes_to_optimize.size())
+        {
+            passes_to_optimize.erase(
+                passes_to_optimize.begin() + num_passes_to_keep_after_pruning.value(),
+                passes_to_optimize.end());
+        }
+
+        // Append our newly generated passes to replace the passes we just removed
+        passes_to_optimize.insert(passes_to_optimize.end(), new_passes.begin(),
+                                  new_passes.end());
```

Comment:
With this implementation passes_to_optimize.size() doesn't necessarily equal num_passes_to_optimize. If this is intentional, maybe rename to specify it is the maximum number of passes to optimize.

Reply:
Sorry, which statement are you referring to here?

Reply:
The number of generated passes is always num_passes_to_optimize.value() - num_passes_to_keep_after_pruning.value()

But the number of passes in passes_to_optimize may be less than num_passes_to_keep_after_pruning.value() because of the merging

So merged + generated <= num_passes_to_optimize

Reply:
Ah, great catch. Changed to:
```
        // Remove the worst paths
        if (num_passes_to_keep_after_pruning.value() < passes_to_optimize.size())
        {
            passes_to_optimize.erase(
                passes_to_optimize.begin() + num_passes_to_keep_after_pruning.value(),
                passes_to_optimize.end());
        }

        // Generate new paths to replace the ones we just removed
        std::vector<Pass> new_passes = generatePasses(
            num_passes_to_optimize.value() - passes_to_optimize.size());

        // Append our newly generated passes to replace the passes we just removed
        passes_to_optimize.insert(passes_to_optimize.end(), new_passes.begin(),
                                  new_passes.end());
```

Reply:
Looks good